src/overview.md

```diff
@@ -0,0 +1,368 @@
+other compilers don't do (e.g. borrow checking) and it has a lot of
+unconventional implementation choices (e.g. queries). We will talk about these
+in turn in this chapter, and in the rest of the guide, we will look at all the
+individual pieces in more detail.
+
+## What the compiler does to your code
+
+So first, let's look at what the compiler does to your code. For now, we will
+avoid mentioning how the compiler implements these steps except as needed;
+we'll talk about that later.
+
+**TODO: Would be great to have a diagram of this once we nail down the details...**
+
+**TODO: someone else should confirm this vvv**
+
+- The compile process begins when a user writes a Rust source program in text
+  and invokes the `rustc` compiler on it. The work that the compiler needs to
+  perform is defined by command-line options. For example, it is possible to
+  enable nightly features (`-Z` flags), perform `check`-only builds, or emit
+  LLVM-IR rather than executable machine code. The `rustc` executable call may
+  be indirect through the use of `cargo`.
+- Command line argument parsing occurs in the [`librustc_driver`]. This crate
+  defines the compile configuration that is requested by the user and passes it
+  to the rest of the compilation process as a [`rustc_interface::Config`].
+- The raw Rust source text is analyzed by a low-level lexer located in
+  [`librustc_lexer`]. At this stage, the source text is turned into a stream of
+  atomic source code units known as _tokens_. (**TODO**: chrissimpkins - Maybe
+  discuss Unicode handling during this stage?)
+- The token stream passes through a higher-level lexer located in
+  [`librustc_parse`] to prepare for the next stage of the compile process. The
+  [`StringReader`] struct is used at this stage to perform a set of validations
+  and turn strings into interned symbols.
+- (**TODO**: chrissimpkins - Expand info on parser) We then [_parse_ the stream
+  of tokens][parser] to build an Abstract Syntax Tree (AST).
+  - macro expansion (**TODO** chrissimpkins)
+  - ast validation (**TODO** chrissimpkins)
+  - nameres (**TODO** chrissimpkins)
+  - early linting (**TODO** chrissimpkins)
+
+- We then take the AST and [convert it to High-Level Intermediate
+  Representation (HIR)][hir]. This is a compiler-friendly representation of the
+  AST.  This involves a lot of desugaring of things like loops and `async fn`.
+- We use the HIR to do [type inference]. This is the process of automatic
+  detection of the type of an expression. **TODO: how `ty` module fits in
+  here**
+- **TODO: Maybe some other things are done here? I think initial type checking
+  happens here? And trait solving?**
+- The HIR is then [lowered to Mid-Level Intermediate Representation (MIR)][mir].
+  - Along the way, we construct the HAIR, which is an even more desugared HIR.
+    HAIR is used for pattern and exhaustiveness checking. It is also more
+    convenient to convert into MIR than HIR is.
```

Comment:
I liked how @eddyb has explained HAIR here https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/history.20of.20typeck.20and.20borrowck. Maybe @eddyb has more thoughts about the Overview :).

Reply:
Saw that down here https://github.com/rust-lang/rustc-dev-guide/pull/633/files#diff-c872a24fac64fbf63e412eb49ccf468aR175 some of the things said in that Zulip topic were included.

Reply:
added to followup issue